File: core/envelope-sim/src/main/java/fr/sncf/osrd/DriverBehaviour.java
package fr.sncf.osrd;

import fr.sncf.osrd.envelope.Envelope;
import fr.sncf.osrd.envelope.MRSPEnvelopeBuilder;
import fr.sncf.osrd.envelope.part.EnvelopePart;
import fr.sncf.osrd.envelope_sim.EnvelopeProfile;

import java.util.List;

public class DriverBehaviour {
    public final double acceleratingPostponementOffset;
    public final double brakingAnticipationOffset;

    public DriverBehaviour() {
        this.acceleratingPostponementOffset = 50;
        this.brakingAnticipationOffset = 100;
    }
    public DriverBehaviour(double acceleratingPostponementOffset, double brakingAnticipationOffset) {
        this.acceleratingPostponementOffset = acceleratingPostponementOffset;
        this.brakingAnticipationOffset = brakingAnticipationOffset;
    }

    public Envelope applyToMRSP(Envelope mrsp) {
        var builder = new MRSPEnvelopeBuilder();
        var totalLength = mrsp.getTotalDistance();
        for(EnvelopePart part: mrsp.getParts()) {
            var begin = part.getBeginPos();
            var end = part.getEndPos();
            // compute driver behaviour offsets
            begin -= this.brakingAnticipationOffset;
            end += this.acceleratingPostponementOffset;
            begin = Math.max(0, begin);
            end = Math.min(totalLength, end);
            var speed = part.getMaxSpeed();

            builder.addPart(EnvelopePart.generateTimes(
                    List.of(EnvelopeProfile.CONSTANT_SPEED, MRSPEnvelopeBuilder.LimitKind.SPEED_LIMIT),
                    new double[]{begin, end},
                    new double[]{speed, speed}
            ));
        }
        return builder.build();
    }
}

File: core/envelope-sim/src/main/kotlin/fr/sncf/osrd/DriverBehaviour.kt
package fr.sncf.osrd

import fr.sncf.osrd.envelope.Envelope
import fr.sncf.osrd.envelope.MRSPEnvelopeBuilder
import fr.sncf.osrd.envelope.part.EnvelopePart
import fr.sncf.osrd.envelope_sim.EnvelopeProfile
import fr.sncf.osrd.path.interfaces.PhysicsPath
import fr.sncf.osrd.utils.OffsetRangeMap
import fr.sncf.osrd.utils.units.Offset
import fr.sncf.osrd.utils.units.meters
import kotlin.math.max
import kotlin.math.min

data class DriverBehaviour(
    val acceleratingPostponementOffset: Double = 50.0,
    val brakingAnticipationOffset: Double = 100.0,
    val signalingSystems: List<String> = listOf("BAL", "BAPR"),
) {
    /** Applies the driver behavior to the MRSP, adding reaction time for MRSP changes */
    fun applyToMRSP(
        mrsp: Envelope,
        optSignalingSystemRanges: OffsetRangeMap<PhysicsPath, String>? = null,
    ): Envelope {
        val signalingSystemRanges = optSignalingSystemRanges ?: OffsetRangeMap()
        val builder = MRSPEnvelopeBuilder()
        val totalLength = mrsp.totalDistance
        for (part in mrsp) {
            var begin = part.beginPos
            var end = part.endPos
            // compute driver behaviour offsets
            if (
                signalingSystems.contains(
                    signalingSystemRanges.get(Offset<PhysicsPath>(begin.meters)) ?: ""
                )
            )
                begin -= this.brakingAnticipationOffset
            if (
                signalingSystems.contains(
                    signalingSystemRanges.get(Offset<PhysicsPath>(end.meters)) ?: ""
                )
            )
                end += this.acceleratingPostponementOffset
            begin = max(0.0, begin)
            end = min(totalLength, end)
            val speed = part.maxSpeed

            builder.addPart(
                EnvelopePart.generateTimes(
                    listOf(
                        EnvelopeProfile.CONSTANT_SPEED,
                        MRSPEnvelopeBuilder.LimitKind.SPEED_LIMIT,
                    ),
                    doubleArrayOf(begin, end),
                    doubleArrayOf(speed, speed),
                )
            )
        }
        return builder.build()
    }
}
